### PR TITLE
Ensure widget is alive before executing code

### PR DESCRIPTION
**Pull request by loco-odoo**
[FIX] sale: ensure widget is alive before executing code

Steps to reproduce the bug:
- Add a delay (for example 10 seconds) at the beginning of the
`/sale/get_combination_info_website` route.
- Go on a product page and enter edit mode. Note that the
`websiteSaleCarouselProduct` is destroyed.
- Click on the the product image. Note that the "Replace" button
appears.
- Wait for the added delay to finish.

-> If you click on the product image, the "Replace" button does not
appear.

In this situation, the added delay represents a slow rpc answer. Let's
analyse the flow of instructions in order to better understand the
problem:
- When you go on a product page, the `WebsiteSale` public widget is
started. `_getCombinationInfo` of `VariantMixin` is then called through
the `start()` of the public widget. As the rpc is taking time to answer,
`_onChangeCombination()` is not directly called.
- When entering edit mode, the `WebsiteSale` public widget is destroyed.
Due to the edit mode, the `o_editable` class has been added on editable
elements.
- After the added delay and thanks to the rpc answer,
`_updateProductImage()` is called through `_onChangeCombination()`. The
`_updateProductImage()` replaces some elements of the DOM and by doing
so, removes the `o_editable` class of some elements. Consequently, the
snippet option linked to the image is not displayed.

The problem here is that the `_updateProductImage` method is called even
if its associated public widget has been destroyed. To solve the
problem, we first check that the associated widget is alive before
handling the result of the rpc answer in the `VariantMixin` mixin.

Note that the first idea was to use `this._rpc()` instead of
`ajax.jsonRpc()` in the mixin. Indeed, the advantage of using
`this._rpc()` is that it already ensures that the associated widget is
alive before handling the result of the rpc answer. The problem is that
some widgets that use the `VariantMixin` are created in such a way that
`this._rpc()` can not be used on them (for example
`OptionalProductsModal`).

Related to runbot-28700

### CODE REVIEW

**Comment by loco-odoo** [comment]
Note:
My fist idea was to write the `_callVariantMixinRpc()` like this:
```js
    /**
     * Call a rpc route with the most appropiate method; either with
     * this._rpc() if it exists and with ajax.jsonRpc() otherwise.
     *
     * @private
     * @param {String} route The route the the rpc has to call.
     * @param {Object} params A dictionary containing the parameter of the rpc.
     */
    _callVariantMixinRpc(route, params) {
        return this._rpc ? this._rpc({route: route, params: params}) : ajax.jsonRpc(route, 'call', params);
    },
```
However, this led to error; for example, it seems that using `this._rpc()` leads to error when its used on the `OptionalProductsModal` component although `this._rpc()` is defined.

**Comment by loco-odoo** [comment]
@qsm-odoo , I have adapted this PR according to the discussion we had yesterday :+1:

**Review by qsm-odoo — APPROVED** [review]
Thanks for the fix, good commit message :+1: 

I'd say we should add a test about this but this is a non critical bug which will be annoying to test (requires to slow down a rpc etc). We'll wait for a better test system for the website. In the end, this is already a source of race condition that breaks tests so... better than nothing. Let's skip.

@robodoo r+

**Comment by loco-odoo** [comment]
Note: waiting for https://github.com/odoo/odoo/pull/149385 to be merged before merging the forward port.
